### PR TITLE
Add priority source to genkw to account for merging customization

### PR DESCRIPTION
**Pull request by xjules**
**Issue**
Resolves https://github.com/equinor/ert/issues/12240

**Approach**
This introduces priority source attribute to genkw to account for custom preference when merging genkw with design matrix.

- [x] add documentation
(Screenshot of new behavior in GUI if applicable)
image

- [x] PR title captures the intent of the changes, and is fitting for release notes.
- [x] Added appropriate release note label
- [x] Commit history is consistent and clean, in line with the [contribution guidelines](https://github.com/equinor/ert/blob/main/CONTRIBUTING.md).
- [x] Make sure unit tests pass locally after every commit (`git rebase -i main
      --exec 'just rapid-tests'`)

## When applicable
- [ ] **When there are user facing changes**: Updated documentation
- [ ] **New behavior or changes to existing untested code**: Ensured that unit tests are added (See [Ground Rules](https://github.com/equinor/ert/blob/main/CONTRIBUTING.md#ground-rules)).
- [ ] **Large PR**: Prepare changes in small commits for more convenient review
- [ ] **Bug fix**: Add regression test for the bug
- [ ] **Bug fix**: Add backport label to latest release (format: 'backport release-branch-name')

### CODE REVIEW

File: docs/ert/getting_started/howto/design_matrix.rst [hunk 1]
```diff
@@ -10,7 +10,7 @@ contains parameters and their values per realization.
 parameters as `DESIGN2PARAMS`.
 
 If the same parameter name is defined both in the design matrix and in the GEN_KW parameter definition,
-the value from the design matrix takes precedence and the parameter will not be updated by ERT.
+the priority of which values are to be used is defined by optional attribute PRIORITY:<design_matrix|sampled>.
```
Comment (flikka) [inline docs/ert/getting_started/howto/design_matrix.rst:13]: It feels confusing that when choosing between "GEN_KW" and "DESIGN_MATRIX" you have to choose between design_matrix and sampled. "sampled" is more a property than a "source".

File: src/ert/config/design_matrix.py [hunk 2]
```diff
@@ -63,13 +68,21 @@ def from_config_list(cls, config_list: list[str | dict[str, str]]) -> DesignMatr
                     "DESIGN_SHEET and DEFAULT_SHEET can not point to the same sheet."
                 ).set_context(config_list)
             )
+        if priority_source not in {"design_matrix", "sampled"}:
+            errors.append(
+                ErrorInfo(
+                    f"PRIORITY must be either 'design_matrix' or 'sampled';"
+                    f" is '{priority_source}'"
```
Comment (flikka) [inline src/ert/config/design_matrix.py]: Would it make sense to use the DataSource enum in gen_kw_config.py? for these strings?

File: src/ert/config/design_matrix.py [hunk 3]
```diff
@@ -47,6 +51,7 @@ def from_config_list(cls, config_list: list[str | dict[str, str]]) -> DesignMatr
         options = cast(dict[str, str], config_list[1])
         design_sheet = options.get("DESIGN_SHEET", "DesignSheet")
         default_sheet = options.get("DEFAULT_SHEET", None)
+        priority_source = options.get("PRIORITY", "design_matrix")
```
Comment (flikka) [inline src/ert/config/design_matrix.py]: use DataSource.DESIGN_MATRIX?

File: src/ert/config/design_matrix.py [hunk 4]
```diff
@@ -154,9 +168,17 @@ def merge_with_existing_parameters(
 
         for param_cfg in existing_parameters:
             if isinstance(param_cfg, GenKwConfig) and param_cfg.name in design_cfgs:
-                param_cfg.input_source = DataSource.DESIGN_MATRIX
-                param_cfg.update = False
-                param_cfg.distribution = RawSettings()
+                param_cfg.input_source = DataSource(
+                    self.parameter_priority.get(param_cfg.name, "design_matrix")
```
Comment (flikka) [inline src/ert/config/design_matrix.py]: Use DataSource.DESIGN_MATRIX as default?

Comment (xjules) [reply src/ert/config/design_matrix.py]: parameter_priority is dict[str,str] and DataSource is a StrEnum so it will convert it correctly. 
But it might work DataSource(DataSource.DESIGN_MATRIX) though...

File: src/ert/config/design_matrix.py (hunk 3 above)
Comment (xjules) [reply src/ert/config/design_matrix.py]: priority_source is a string.

File: docs/ert/getting_started/howto/design_matrix.rst (hunk 1 above)
Comment (jonathan-eq) [inline docs/ert/getting_started/howto/design_matrix.rst:13]: It is a bit unclear what will happen if you have the same parameter in multiple design matrices and gen kw. If the first DM has PRIORITY:sampled and the second one has PRIORITY:design_matrix, will it use design_matrix? Also, what will happen if that order is swapped?

File: docs/ert/reference/configuration/keywords.rst [hunk 5]
```diff
@@ -320,20 +325,35 @@ wherein coeff_priors
...
 
-So the genkw would keep its name (COEFFS) and the design matrix parameters would be prefixed with DESIGN_MATRIX.
-Note that GEN_KW parameters that are not found in the design matrix will be sampled normally (eg. COEFFS:f).
+So the design matrix parameters would be without the prefix.
+
+In case we specify sampled priority:
+::
+
+        GEN_KW COEFFS coeff_priors
+        DESIGN_MATRIX poly_design.xlsl DESIGN_SHEET:DesignSheet DEFAULT_SHEET:DefaultSheet PRIORITY:sampled
```
Comment (jonathan-eq) [inline docs/ert/reference/configuration/keywords.rst]: Could this be 
```
 GEN_KW COEFFS coeff_priors PRIORITY:5
 DESIGN_MATRIX poly_design.xlsx DESIGN_SHEET:DesignSheet DEFAULT_SHEET:DefaultSheet PRIORITY:3
 DESIGN:MATRIX poly_design2.xlsx DESIGN_SHEET:DesignSheet DEFAULT_SHEET:DefaultSheet PRIORITY:1
```

Comment (jonathan-eq) [inline docs/ert/reference/configuration/keywords.rst]:
```suggestion
        GEN_KW COEFFS coeff_priors
        DESIGN_MATRIX poly_design.xlsx DESIGN_SHEET:DesignSheet DEFAULT_SHEET:DefaultSheet PRIORITY:sampled
```

File: src/ert/config/design_matrix.py (hunk 3 above)
Comment (jonathan-eq) [reply src/ert/config/design_matrix.py]: Is DataSource a StrEnum?

File: src/ert/config/design_matrix.py (hunk 2 above)
Comment (jonathan-eq) [reply src/ert/config/design_matrix.py]: Yes, maybe this would be good!

File: src/ert/config/design_matrix.py [hunk 6]
```diff
@@ -133,6 +146,7 @@ def merge_with_other(self, dm_other: DesignMatrix) -> None:
             for cfg in dm_other.parameter_configurations
             if cfg.name not in common_keys
         )
+        self.parameter_priority.update(dm_other.parameter_priority)
```
Comment (jonathan-eq) [inline src/ert/config/design_matrix.py:152]: If there are multiple design_matrices with overlapping parameters and different PRIORITY options, this will silently overwrite it. Is that something we should warn about?

Comment (jonathan-eq) [reply src/ert/config/design_matrix.py:152]: Or if even one of them has PRIORITY:sampled, should we prioritise that?

File: tests/ert/unit_tests/config/test_analysis_config.py [hunk 7]
```diff
@@ -176,6 +177,28 @@ def test_invalid_design_matrix_format_raises_validation_error():
...
+                        {
+                            "DESIGN_SHEET": "sheet1",
+                            "DEFAULT_SHEET": "sheet2",
+                            "PRIORITY": "invalid_value",
+                        },
+                    ],
+                ],
+            }
+        )
+
```
[the remark this answers is not in the document]
Comment (xjules) [reply tests/ert/unit_tests/config/test_analysis_config.py:194]: Not sure if we need these tests.

File: src/ert/run_models/_create_run_path.py [hunk 8]
```diff
@@ -54,7 +55,12 @@ def _value_export_txt(
         for key, param_map in values.items():
             for param, value in param_map.items():
                 if isinstance(value, (int | float)):
-                    print(f"{key}:{param} {value:g}", file=f)
+                    if key == DESIGN_MATRIX_GROUP:
+                        print(f"{param} {value:g}", file=f)
+                    else:
+                        print(f"{key}:{param} {value:g}", file=f)
+                elif key == DESIGN_MATRIX_GROUP:
+                    print(f"{param} {value}", file=f)
```
Comment (jonathan-eq) [inline src/ert/run_models/_create_run_path.py:63]: Why do we have a different format on this print compared to the one on line 59? What is :g format?

**Review by jonathan-eq — COMMENTED** [review]
Some nitpicking :)

File: tests/ert/ui_tests/cli/analysis/test_design_matrix.py [hunk 9]
```diff
@@ -389,15 +391,15 @@ def _evaluate(coeffs, x):
             ensemble = experiment.get_ensemble_by_name(f"{ensemble_name}{i}")
 
             # coeffs_a should be different in all realizations
```
Comment (jonathan-eq) [inline tests/ert/ui_tests/cli/analysis/test_design_matrix.py:393]: Is this comment still valid?

File: tests/ert/unit_tests/sensitivity_analysis/test_design_matrix.py [hunk 10]
```diff
@@ -80,13 +80,21 @@ def test_merge_multiple_occurrences(
...
+    }
     _create_design_matrix(
         tmp_path / "design_matrix_2.xlsx", design_sheet_pd, default_sheet_pd
     )
     design_matrix_2 = DesignMatrix(
-        tmp_path / "design_matrix_2.xlsx", "DesignSheet", "DefaultSheet"
+        tmp_path / "design_matrix_2.xlsx",
+        "DesignSheet",
+        "DefaultSheet",
+        "design_matrix",
```
Comment (jonathan-eq) [inline tests/ert/unit_tests/sensitivity_analysis/test_design_matrix.py:84]: Can we add `assert design_matrix_1.priority_source == "design_matrix"`?

File: tests/ert/unit_tests/sensitivity_analysis/test_design_matrix.py [hunk 11]
```diff
@@ -103,20 +111,100 @@ def test_merge_multiple_occurrences(
...
+@pytest.mark.parametrize(
+    "parameters, priority, num_configs, input_source, group_name",
+    [
+        pytest.param(
+            ["a", "b"],
+            "design_matrix",
+            2,
+            {"a": DataSource.DESIGN_MATRIX, "b": DataSource.DESIGN_MATRIX},
+            {"a": DESIGN_MATRIX_GROUP, "b": DESIGN_MATRIX_GROUP},
+            id="genkw_replaced_priority_design_matrix",
```
Comment (jonathan-eq) [inline tests/ert/unit_tests/sensitivity_analysis/test_design_matrix.py]: Does this mean genkw took priority or did design_matrix take priority?

Comment (xjules) [reply tests/ert/unit_tests/sensitivity_analysis/test_design_matrix.py]: This means that design_matrix has priority

File: docs/ert/getting_started/howto/design_matrix.rst (hunk 1 above)
Comment (xjules) [reply docs/ert/getting_started/howto/design_matrix.rst:13]: The latter one takes precedence

File: docs/ert/reference/configuration/keywords.rst (hunk 5 above)
Comment (xjules) [reply docs/ert/reference/configuration/keywords.rst]: I think it is enough to have priorities on DM. We thought about adding it to prior file, but that would be too much effort and cumbersome to maintain for the users.

Comment (xjules) [reply docs/ert/reference/configuration/keywords.rst]: Good catch! There are many wrong occurrences :)

File: tests/ert/unit_tests/sensitivity_analysis/test_design_matrix.py (hunk 11 above)
Comment (jonathan-eq) [reply tests/ert/unit_tests/sensitivity_analysis/test_design_matrix.py]: Can we add a double underscore to make this more clear?
```
 id="genkw_replaced__priority_design_matrix"
```

Comment (xjules) [reply tests/ert/unit_tests/sensitivity_analysis/test_design_matrix.py]: maybe: `overlap_priority_design_matrix`?

File: src/ert/config/design_matrix.py (hunk 3 above)
Comment (xjules) [reply src/ert/config/design_matrix.py]: yes. Maybe I can do `DataSource.DESIGN_MATRIX.value`?

File: src/ert/config/design_matrix.py (hunk 2 above)
Comment (xjules) [reply src/ert/config/design_matrix.py]: Oh, this is meant for the actual error message?

File: src/ert/config/design_matrix.py (hunk 6 above)
Comment (xjules) [reply src/ert/config/design_matrix.py:152]: That's very much a corner case. If we are to solve it - one possibility is to just validate that all `DESIGN_MATRIX` instances have the same priority setting.

File: tests/ert/unit_tests/sensitivity_analysis/test_design_matrix.py (hunk 10 above)
Comment (xjules) [reply tests/ert/unit_tests/sensitivity_analysis/test_design_matrix.py:84]: I guess you mean:
`assert design_matrix_2.priority_source == "design_matrix"`

File: tests/ert/unit_tests/sensitivity_analysis/test_design_matrix.py (hunk 11 above)
Comment (xjules) [reply tests/ert/unit_tests/sensitivity_analysis/test_design_matrix.py]: changed

File: tests/ert/ui_tests/cli/analysis/test_design_matrix.py (hunk 9 above)
Comment (xjules) [reply tests/ert/ui_tests/cli/analysis/test_design_matrix.py:393]: we need to write an issue to refactor this test

File: src/ert/config/design_matrix.py (hunk 6 above)
Comment (xjules) [reply src/ert/config/design_matrix.py:152]: this has been resolved in https://github.com/equinor/ert/commit/2f0b15a03b7c8a0614cfe3e6a4d9a6411758605a

File: src/ert/run_models/_create_run_path.py (hunk 8 above)
Comment (xjules) [reply src/ert/run_models/_create_run_path.py:63]: That is also taking care of the scientific notations; ie. 1000 -> 1e4

File: docs/ert/reference/configuration/keywords.rst [hunk 12]
```diff
@@ -320,20 +325,35 @@ wherein coeff_priors
...
+
+the final set of parameters (for example in parameters.txt in real==0) would be:
+::
+
+        a=1 <- design value
+        COEFFS:b=0.2 <- sampled value
+        COEFFS:c=1.3 <- sampled value
+        COEFFS:d=7.8 <- sampled value
+        e=1 <- design value
+        COEFFS:f=5.7 <- sampled value
```
Comment (jonathan-eq) [inline docs/ert/reference/configuration/keywords.rst:356]: Very good examples of both variants ))

File: tests/ert/unit_tests/sensitivity_analysis/test_design_matrix.py (hunk 10 above)
Comment (jonathan-eq) [reply tests/ert/unit_tests/sensitivity_analysis/test_design_matrix.py:84]: Very good!

File: src/ert/run_models/_create_run_path.py (hunk 8 above)
Comment (jonathan-eq) [inline src/ert/run_models/_create_run_path.py:63]: Should we add tests for this?

Comment (xjules) [reply src/ert/run_models/_create_run_path.py:63]: I thought about that, but basically we don't have any tests for parameters.txt (we don't use it). So not sure

Comment (jonathan-eq) [reply src/ert/run_models/_create_run_path.py:63]: Sounds like a problem for the future, maybe GoodFirstIssue :D

**Review by jonathan-eq — APPROVED** [review]
Straight to main!
